### PR TITLE
fix: Update and fix e2e tests to handle the presence of the offscreen document under MV3

### LINKED ISSUE #21496: Fix and create a success e2e tests pipeline against mv3 build
- Filtering out tests are flakier in MV3 than MV2, 
- Not necessary need to fix flaky ones overlapping

### PR DESCRIPTION
**Pull request by danjm**
## **Description**

Our MV3 builds add a "Offscreen Document" to our extension. This means that when our e2e tests call `driver.getAllWindowHandles()`, a new window is included in the list (the window for the offscreen document). This PR updates the e2e tests to handle this change. Comments are left on each file to explain why the change was needed.

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/24675?quickstart=1)

## **Related issues**

Fixes: https://github.com/MetaMask/metamask-extension/issues/21496

## **Manual testing steps**

Any of the test files modified in this PR should pass when run locally with MV3 builds

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: test/e2e/accounts/common.ts [hunk 1]
```diff
@@ -182,7 +183,20 @@ async function switchToAccount2(driver: Driver) {
 }
 
 export async function connectAccountToTestDapp(driver: Driver) {
```
Comment (danjm) [inline test/e2e/accounts/common.ts:185]: `switchToOrOpenDapp` attempts "an unusually fast switchToWindowWithTitle, just 1 second" and if that fails, then calls `openDapp`. The problem the `switchToWindowWithTitle` assumes that there will be a window other than the offscreen document present. It seems that `openDapp` fails if the drivers current window is the offscreen document. The change in this file ensures that, for this e2e test, the extension window is selected first and then the dapp can be opened from there.

File: test/e2e/helpers.js [hunk 2]
```diff
@@ -655,7 +655,7 @@ const DAPP_URL = 'http://127.0.0.1:8080';
 const DAPP_ONE_URL = 'http://127.0.0.1:8081';
 
 const openDapp = async (driver, contract = null, dappURL = DAPP_URL) => {
```
Comment (danjm) [inline test/e2e/helpers.js:657]: This is needed for the update to test/e2e/json-rpc/switchEthereumChain.spec.js, which can be seen below

File: test/e2e/json-rpc/switchEthereumChain.spec.js [hunk 3]
```diff
@@ -53,13 +53,8 @@ describe('Switch Ethereum Chain for two dapps', function () {
         await driver.clickElement('.request-queue-toggle');
```
Comment (danjm) [inline test/e2e/json-rpc/switchEthereumChain.spec.js:54]: This file assumed that window handles at certain indices would always map to the dapps that are opened. The problem is that the offscreen document can actually be one of the window handles at those indices, The solution is to return the new handle with the `openDapp` call, and then use that handle to select the right windows for the opened dapps

File: test/e2e/tests/settings/full-size-view-settings.spec.js [hunk 4]
```diff
@@ -33,28 +33,25 @@ describe('Full-size View Setting @no-mmi', function () {
         await unlockWallet(driver);
         await toggleFullSizeViewSetting(driver);
         await openDapp(driver);
```
Comment (danjm) [inline test/e2e/tests/settings/full-size-view-settings.spec.js:35]: This file assumed that window handles at certain indices would always map to the dapp and popup window, but again the offscreen document can be at one of these indices.

The solution, which maintains what this test was actually testing, is to update it to compare window handles before and after the user action being tested, and to make an assertion against the new window handle

File: test/e2e/webdriver/driver.js [hunk 5]
```diff
@@ -509,8 +509,9 @@ class Driver {
   }
 
   async openNewPage(url) {
```
Comment (danjm) [inline test/e2e/webdriver/driver.js:511]: This change is needed to support the changes in switchEthereumChain.spec.js (which can be seen above)

File: test/e2e/webdriver/driver.js [hunk 6]
```diff
@@ -534,12 +535,12 @@ class Driver {
     return await this.driver.getAllWindowHandles();
   }
 
-  async waitUntilXWindowHandles(x, delayStep = 1000, timeout = this.timeout) {
```
Comment (danjm) [inline test/e2e/webdriver/driver.js:537]: Under MV3, there will always be 1 more window handle than previously assumed, so this is a global fix for `waitUntilXWindowHandles`, to accomodate the MV3 offscreen document

**Review by DDDDDanica — APPROVED** [review]

**Comment by danjm** [comment]
The `test-e2e-chrome` failure here is related to changes in this PR in the `request-queue` test. I thought that change would be safe for mv2... I will investigate

**Review by vthomas13 — APPROVED** [review]

**Comment by danjm** [comment]
> The test-e2e-chrome failure here is related to changes in this PR in the request-queue test. I thought that change would be safe for mv2... I will investigate

I was unable to repro this locally, and it didn't occur again after multiple re-runs on CI (two on this branch and two on another branch I pushed up).

There probably is some flakiness here, but it is hard to recreate, even on CI. I am going to merge and then we can observe if this flake happens again, as we will have this test running more frequently once merge.

**Pull request merged**